File: 0016 最接近的三数之和/main.py
from typing import List
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        ans = nums[0]+nums[1]+nums[2]
        for i in range(len(nums)-2):
            li = i+1
            ri = len(nums)-1
            while(li != ri):
                tmp = nums[li] + nums[ri]+nums[i]
                if(abs(tmp-target) < abs(ans-target)):
                    ans = tmp
                if(tmp-target < 0):
                    li += 1
                elif(tmp-target > 0):
                    ri -= 1
                else:
                    return ans
        return ans
a = Solution()
a.threeSumClosest([1, 2, 4, 8, 16, 32, 64, 128], 82)

File: 0016 最接近的三数之和/main.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int li,ri;
        int numli,numri;
        int res=INT_MAX;
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size()-2;i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }
            li=i+1;
            ri=nums.size()-1;
            while(ri>li){
                numli = nums[li];
                numri = nums[ri];
                int sum=numri + numli + nums[i];
                if( abs( sum - target) < abs( res  -target ) ){
                    res=sum;
                }
                if( sum > target){
                    ri--;
                }else{
                    li++;
                }
            }
        }
        return res;
    }
};